governance/microservice.go

```diff
@@ -0,0 +1,485 @@
+		// get next lower version
+		if new_msdef, err := microservice.GetRollbackMicroserviceDef(w.exchHandlers.GetHTTPMicroserviceHandler(), msdef, w.deviceId, w.deviceToken, w.db); err != nil {
+			return fmt.Errorf(logString(fmt.Sprintf("Error finding the new microservice definition to downgrade to for %v %v version key %v. error: %v", msdef.SpecRef, msdef.Version, msdef.Id, err)))
+		} else if new_msdef == nil { //no more to try, exit out
+			return fmt.Errorf(logString(fmt.Sprintf("Unable to find the microservice definition to downgrade to for %v %v version key %v. error: %v", msdef.SpecRef, msdef.Version, msdef.Id, err)))
+		} else if err := w.UpgradeMicroservice(msdef, new_msdef, false); err != nil {
+			glog.Errorf(logString(fmt.Sprintf("Failed to downgrade %v from version %v key %v to version %v key %v. %v", msdef.SpecRef, msdef.Version, msdef.Id, new_msdef.Version, new_msdef.Id, err)))
+		} else {
+			return nil
+		}
+	}
+
+	return nil
+}
+
+// start an microservice instance for the given agreement according to the microservice sharing mode
+func (w *GovernanceWorker) startMicroserviceInstForAgreement(msdef *persistence.MicroserviceDefinition, agreementId string, protocol string) error {
+	glog.V(3).Infof(logString(fmt.Sprintf("start microserivce instance %v for agreement %v", msdef.SpecRef, agreementId)))
+
+	var msi *persistence.MicroserviceInstance
+	needs_new_ms := false
+
+	// always start a new ms instance if the sharing mode is multiple
+	if msdef.Sharable == exchange.MS_SHARING_MODE_MULTIPLE {
+		needs_new_ms = true
+		// for other sharing mode, start a new ms instance only if there is no existing one
+		// the "exclusive" sharing mode is taken cared by the maxAgreements=1 so that there will no be more than one agreements to come at any time
+	} else if ms_insts, err := persistence.FindMicroserviceInstances(w.db, []persistence.MIFilter{persistence.AllInstancesMIFilter(msdef.SpecRef, msdef.Version), persistence.UnarchivedMIFilter()}); err != nil {
+		return fmt.Errorf(logString(fmt.Sprintf("Error retrieving all the microservice instaces from db for %v version %v key %v. %v", msdef.SpecRef, msdef.Version, msdef.Id, err)))
+	} else if ms_insts == nil || len(ms_insts) == 0 {
+		needs_new_ms = true
+	} else {
+		msi = &ms_insts[0]
+	}
+
+	if needs_new_ms {
+		var inst_err error
+		if msi, inst_err = w.StartMicroservice(msdef.Id); inst_err != nil {
+
+			// try to downgrade the microservice to a lower version
+			glog.V(3).Infof("Ending the agreement: %v because microservice %v failed to start", agreementId, msdef.SpecRef)
+			ag_reason_code := w.producerPH[protocol].GetTerminationCode(producer.TERM_REASON_MS_DOWNGRADE_REQUIRED)
+			ag_reason_text := w.producerPH[protocol].GetTerminationReason(ag_reason_code)
+			w.cancelAgreement(agreementId, protocol, ag_reason_code, ag_reason_text)
+
+			glog.V(3).Infof("Downgrading microservice %v because version %v key %v failed to start. Error: %v", msdef.SpecRef, msdef.Version, msdef.Id, inst_err)
```

Comment:
Here's another missing use of logString. Please  check all new code for this.

Reply:
ok